### PR TITLE
Detect a missing key error and print a cleaner message.

### PR DESCRIPTION
**Pull request by brendandburns**
Closes https://github.com/kubernetes/kubernetes/issues/63247

@kubernetes/sig-cli-bugs

### CODE REVIEW

**Review by adohe-zz — APPROVED** [review]

File: pkg/kubectl/cmd/patch.go [hunk 1]
```diff
@@ -306,7 +306,15 @@ func getPatchedJSON(patchType types.PatchType, originalJS, patchJS []byte, gvk s
 		if err != nil {
 			return nil, err
 		}
-		return patchObj.Apply(originalJS)
+		bytes, err := patchObj.Apply(originalJS)
+		// TODO: This is pretty hacky, we need a better structured error from the json-patch
+		if err != nil && strings.Contains(err.Error(), "doc is missing key") {
+			msg := err.Error()
+			ix := strings.Index(msg, "key:")
+			key := msg[ix+5:]
```
[the remark this answers is not in the document]
Comment (php-coder) [reply pkg/kubectl/cmd/patch.go]: > Is `key:` a prefix?

As far I can see, based on the original issue, it's something like a separator between useless part and needed one:

> error: jsonpatch replace operation does not apply: **doc is missing key**: /spec/template/spec/containers/0/image

> If so, using `strings.TrimPrefix` is better?

I don't think so, given that the prefix will be long and could change over time.

File: pkg/kubectl/cmd/patch.go [hunk 2]
```diff
@@ -306,7 +306,15 @@ func getPatchedJSON(patchType types.PatchType, originalJS, patchJS []byte, gvk s
 		if err != nil {
 			return nil, err
 		}
-		return patchObj.Apply(originalJS)
+		bytes, err := patchObj.Apply(originalJS)
+		// TODO: This is pretty hacky, we need a better structured error from the json-patch
+		if err != nil && strings.Contains(err.Error(), "doc is missing key") {
+			msg := err.Error()
```
Comment (php-coder) [inline pkg/kubectl/cmd/patch.go]: Possible tiny improvement:
`msg` variable could be moved out of `if`, so we'll be able to use it inside of `strings.Contains()`.

File: pkg/kubectl/cmd/patch.go [hunk 3]
```diff
@@ -306,7 +306,15 @@ func getPatchedJSON(patchType types.PatchType, originalJS, patchJS []byte, gvk s
 		if err != nil {
 			return nil, err
 		}
-		return patchObj.Apply(originalJS)
+		bytes, err := patchObj.Apply(originalJS)
+		// TODO: This is pretty hacky, we need a better structured error from the json-patch
+		if err != nil && strings.Contains(err.Error(), "doc is missing key") {
```
Comment (php-coder) [inline pkg/kubectl/cmd/patch.go]: Perhaps, it will better if we would search for `doc is missing key:` (with a trailing colon).

**Pull request merged**